### PR TITLE
feat: wire-up Ledger hdwallet

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Working in this base functionality PR:

- [x] Receive
- [x] Sends
- [x] Txs
- [x] Swaps - with caveats, only same-chain for now until we cache things properly
- [x] Re-pair device / race `'TransportRaceCondition: An action was already pending on the Ledger device. Please deny or reconnect.'` condition
- [x] Reconnect on app close disconnect i.e `DisconnectedDeviceDuringOperation: Failed to execute 'transferOut' on 'USBDevice': The device was disconnected.`

To be implemented in follow-ups:

- Ledger support for Cosmos and THORChain (RUNE)
- BCH receive address format - fixed in https://github.com/shapeshift/hdwallet/pull/640
- Accounts caching and/or UX improvements WRT the need to open xyz app first
- Verification of addresses at key times so users can ensure they're dealing with the right Ledger/account
- Programatically open the relevant app before `call()` to specific apps?
- Cross-chain swaps - can't be implemented until either/all 3 above are implemented
## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Risk

Large regression risk in `useReceiveAddress` / `getReceiveAddress` being too strongly memoized i.e stale receive address on wallet change/"update" for all wallets. Think things like next receive address being broken for UTXOs, receive addresses when switching wallets or MM account being the one from the previously connected wallet/account, etc.

## Testing

- Ledger pairing works
- Ledger chains connection works (currently requires to open the matching app for each chain before clicking "Connect" or doing any operation i.e Tx signing or receive address for a given chain
- Sends work for all chains (assumes the relevant app is currently open before starting the send flow)
  - Make sure to test with Legacy/SegWit/SegWit Native for LTC and BTC
- Accounts are loaded for all chains and script types i.e all EVM chains, DOGE, BCH, BTC (all 3 script types) and LTC (all 3 script types)
- Ensure the accounts are matching the ones from native - import the same seed as native into your Ledger for testing

### Engineering

- All Ledger capability is safely flagged off

### Operations

- ☝🏽 

## Screenshots (if applicable)

image
image
image

https://github.com/shapeshift/web/assets/17035424/f70c59d0-6ed1-450d-9913-cd31a56e9060

image
image
image
image
image
image
image
image

Note the non-standard receive address for BCH here - to be tackled in a follow-up. It does work (tested with self sends as can be seen above), although does shows up with a warning WRT unverfied inputs on the Ledger currently, which may be related to the address format.

image
image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5336** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5336" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5336?utm_source=stack-comment).

File: src/components/Icons/LedgerIcon.tsx
Comment (gomesalexandre) [inline src/components/Icons/LedgerIcon.tsx:3]:
```suggestion
 * Ledger Icon
```

File: src/components/Icons/LedgerIcon.tsx
Comment (gomesalexandre) [inline src/components/Icons/LedgerIcon.tsx:5]:
```suggestion
 * @see https://developers.ledger.com/docs/transport/ledger-logos/
```

File: src/components/MultiHopTrade/hooks/useReceiveAddress.tsx
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useReceiveAddress.tsx:35]: This improves perf a *ton* and doesn't re-run this guy so many times - which rugged Ledger as every `wallet.getAddress()` call is an effective device prompt and requires the Ledger to be connected on the right app for that call.

This has regression risks, as we're not invalidating on `wallet` change, since wallet isn't serializable. Perhaps `deviceId` as a property is something we could use to at least invalidate on `deviceId` change?

File: src/context/WalletProvider/Ledger/components/Chains.tsx
Comment (gomesalexandre) [inline src/context/WalletProvider/Ledger/components/Chains.tsx:69]: Will bring unique `ledger-<uuid>` as a follow-up

File: src/context/WalletProvider/Ledger/components/Connect.tsx
Comment (gomesalexandre) [inline src/context/WalletProvider/Ledger/components/Connect.tsx:54]: Will bring `ledger-<uuid>` as a follow-up

File: .yarnrc.yml
Comment (gomesalexandre) [inline .yarnrc.yml:19]: See this failed run https://github.com/shapeshift/web/actions/runs/6349258371/job/17247184345

Will need to re-publish all snaps packages and bump hdwallet adapter to consume these

File: src/components/MultiHopTrade/hooks/useReceiveAddress.tsx
Comment (gomesalexandre) [reply src/components/MultiHopTrade/hooks/useReceiveAddress.tsx]: https://github.com/shapeshift/web/pull/5336/commits/18c73c57c643efcbc5b7aeed881c0a17297d47c9

File: src/components/MultiHopTrade/hooks/useReceiveAddress.tsx
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useReceiveAddress.tsx:17]: This ran a *lot* more than necessary and was causing a bunch of spam (to the Ledger, but not only, this runs way too much regardless of the wallet) at best, but more often than not, actual bugs - think trying to run `useReceiveAddress` while having the wrong app opened

Comment (gomesalexandre) [reply src/components/MultiHopTrade/hooks/useReceiveAddress.tsx:17]: Note:
1. This is now p-memoized, meaning rejections won't be cached so we can attempt again in case of a failed call and aren't rugged for the lifetime of the sesh
2. This is now exported for pure use outside of react hooks

File: src/components/MultiHopTrade/hooks/useReceiveAddress.tsx
Comment (gomesalexandre) [inline src/components/MultiHopTrade/hooks/useReceiveAddress.tsx:30]: This would rug the p-memoization and never throw, meaning we would cache failures as `undefined`

File: src/context/WalletProvider/Ledger/components/Connect.tsx
Comment (gomesalexandre) [reply src/context/WalletProvider/Ledger/components/Connect.tsx:54]: Or perhaps not - given they all use `0001` as a serial number, the uuid trick won't cut it either as it would generate a new deviceId every connection. We'll have to go with the fact we can't identify a given Ledger and build around that UX-wise.
`ledger-0001` would be it

File: src/context/WalletProvider/WalletProvider.tsx
Comment (woodenfurniture) [inline src/context/WalletProvider/WalletProvider.tsx:782]: 🧠

File: src/context/WalletProvider/SelectModal.tsx
Comment (woodenfurniture) [inline src/context/WalletProvider/SelectModal.tsx:57]: there's also `useFeatureFlag` for these use cases

File: src/context/WalletProvider/Ledger/components/Connect.tsx
Comment (woodenfurniture) [inline src/context/WalletProvider/Ledger/components/Connect.tsx:81]:
```suggestion
    />
```

File: src/context/WalletProvider/Ledger/components/Connect.tsx
Comment (woodenfurniture) [inline src/context/WalletProvider/Ledger/components/Connect.tsx:29]: can do a `useCallback` here

File: src/context/WalletProvider/Ledger/components/Connect.tsx
Comment (woodenfurniture) [inline src/context/WalletProvider/Ledger/components/Connect.tsx:31]: do we want to handle the case where pairing is already in progress, i.e `isLoading === true` ?

File: src/context/WalletProvider/Ledger/components/Chains.tsx
Comment (woodenfurniture) [inline src/context/WalletProvider/Ledger/components/Chains.tsx:138]: would it be worth memoizing the sorting of availableAssets?

File: src/context/WalletProvider/Ledger/components/Chains.tsx
Comment (woodenfurniture) [inline src/context/WalletProvider/Ledger/components/Chains.tsx:154]: not sure how fussy we want to be on performance here, but if its worth it you can break this out into a mini component

File: src/context/WalletProvider/Ledger/components/Chains.tsx
Comment (woodenfurniture) [inline src/context/WalletProvider/Ledger/components/Chains.tsx:82]: do we want logging for failed requests here?

File: src/components/MultiHopTrade/types.ts
Comment (woodenfurniture) [inline src/components/MultiHopTrade/types.ts:49]: ❤️

File: src/components/MultiHopTrade/hooks/useReceiveAddress.tsx
Comment (woodenfurniture) [inline src/components/MultiHopTrade/hooks/useReceiveAddress.tsx:33]: would be best to use stable stringification here, else stringify an array (since js objects are specced to be unordered {even though ackchyually they are stable})

File: src/components/MultiHopTrade/hooks/useReceiveAddress.tsx
Comment (woodenfurniture) [inline src/components/MultiHopTrade/hooks/useReceiveAddress.tsx:35]: the ordering of this and `accountMetadata` keys arent guaranteed

File: src/components/MultiHopTrade/hooks/useReceiveAddress.tsx
Comment (woodenfurniture) [reply src/components/MultiHopTrade/hooks/useReceiveAddress.tsx:17]: 🧠

File: src/components/Modals/Receive/ReceiveInfo.tsx
Comment (woodenfurniture) [inline src/components/Modals/Receive/ReceiveInfo.tsx:75]: not introduced in this pr, but do we want loading state for this?

**Review by woodenfurniture — APPROVED** [review]
conceptually LGTM, pushing this to wood for ops to test before merging (as i don't have a ledger)

File: src/context/WalletProvider/SelectModal.tsx
Comment (gomesalexandre) [reply src/context/WalletProvider/SelectModal.tsx]: All the ways to use feature flags! fdb05a09638689ba97ba0900cc134cef99bc428c

File: src/context/WalletProvider/Ledger/components/Connect.tsx
Comment (gomesalexandre) [reply src/context/WalletProvider/Ledger/components/Connect.tsx:29]: Abso-lootely b68309d714

File: src/context/WalletProvider/Ledger/components/Connect.tsx
Comment (gomesalexandre) [reply src/context/WalletProvider/Ledger/components/Connect.tsx:33]: Don't believe that can be the case? Since this is triggered by pair button click, and as the button goes in loading state, it will be disabled

File: src/context/WalletProvider/Ledger/components/Chains.tsx
Comment (gomesalexandre) [reply src/context/WalletProvider/Ledger/components/Chains.tsx]: We absolutely may - 9f7a6191

File: src/components/Modals/Receive/ReceiveInfo.tsx
Comment (gomesalexandre) [reply src/components/Modals/Receive/ReceiveInfo.tsx:75]: We have a skeleton on l.163 which acts as a loading state on `Boolean(receiveAddress)` so I believe we don't need an additional field! https://github.com/shapeshift/web/blob/9f7a61918ad443553504edd2f7e6a374f316347b/src/components/Modals/Receive/ReceiveInfo.tsx#L163

File: src/context/WalletProvider/Ledger/components/Chains.tsx
Comment (gomesalexandre) [reply src/context/WalletProvider/Ledger/components/Chains.tsx]: 8521dd1d5529399074082153de1460e46bcd3b42

File: src/components/MultiHopTrade/hooks/useReceiveAddress.tsx
Comment (gomesalexandre) [reply src/components/MultiHopTrade/hooks/useReceiveAddress.tsx:36]: f0ddfd35137f84ed46a609082dfcc071d28e5aa4

File: src/components/MultiHopTrade/hooks/useReceiveAddress.tsx
Comment (gomesalexandre) [reply src/components/MultiHopTrade/hooks/useReceiveAddress.tsx]: f0ddfd35137f84ed46a609082dfcc071d28e5aa4

File: src/context/WalletProvider/Ledger/components/Chains.tsx
Comment (gomesalexandre) [reply src/context/WalletProvider/Ledger/components/Chains.tsx]: how about this 5039c7754fa1e6741de3e0372f952cb7d7a0b54

**Comment by gomesalexandre** [comment]
Merging as this has ops signoff

**Pull request merged**
